src/helpers.cpp

```diff
@@ -94,27 +94,43 @@ namespace ada::helpers {
     return pos > input.size() ? std::string_view() : input.substr(pos);
   }
 
-  ada_really_inline size_t get_host_delimiter_location(const ada::url& url, std::string_view& view, bool& inside_brackets) noexcept {
-    size_t location = url.is_special() ? view.find_first_of(":[/?\\") : view.find_first_of(":[/?");
-
-    // Next while loop is almost never taken!
-    while((location != std::string_view::npos) && (view[location] == '[')) {
-      location = view.find(']',location);
-      if(location == std::string_view::npos) {
-        inside_brackets = true;
-        /**
-         * TODO: Ok. So if we arrive here then view has an unclosed [,
-         * Is the URL valid???
-         */
-      } else {
-        location = url.is_special() ? view.find_first_of(":[/?\\#", location) : view.find_first_of(":[/?#", location);
+  ada_really_inline std::pair<size_t,bool> get_host_delimiter_location(const bool is_special, std::string_view& view) noexcept {
+    const size_t view_size = view.size();
+    size_t location = 0;
+    bool found_colon = false;
+    // It is marginally better to have two loops.
+    if(is_special) {
+      for (;location < view_size; ++location) {
+        if (view[location] == '[') {
+          location = view.find(']', location);
+          if (location == std::string_view::npos) {
+            break;
```

Comment:
I don't like how we removed `inside_brackets` here. We need to make sure we are spec compliant according to the only place where it mentions `insideBrackets`:

https://url.spec.whatwg.org/#hostname-state

Reply:
So let us discuss this. The only place you use the `inside_brackets` variable is to check whether ':' was found, we never use it for anything else. But that is wasteful. Why would you not directly record whether ':' was found?

Look at what the existing code does... It calls a function and fills in this boolean variable `inside_brackets`, check whether ':' was found and so forth. Then it exists, and then it does this complicated check, which includes `inside_brackets` to check whether ':' was found. 

But that's unnecessary both conceptually and from a performance point of view: we know, within the subfunction, whether ':' is found, so let us return that.

Reply:
The spec describes an implementation, we have stopped following that quite a long time ago.

Here is how I read the spec, it says...

```
Otherwise, if [c](https://url.spec.whatwg.org/#c) is U+003A (:) and insideBrackets is false, then:
```

That's an unfortunate way to put it but what it asks is whether ':' was found while not inside brackets.

I will add a comment in my code.

Reply:
I see. `found_colon` is false by default and `break` statement inside this just makes sure that `c is : and insideBrackets is false`.

Reply:
Right but I will be adding explanations.